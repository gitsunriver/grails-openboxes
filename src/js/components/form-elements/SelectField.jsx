import _ from 'lodash';
import React from 'react';
import Select from 'react-select';

import 'react-select/dist/react-select.css';

import BaseReactField from './BaseReactField';

const SelectField = (props) => {
  const renderInput = (attributes) => {
    const options = _.map(attributes.options, value => (typeof value === 'string' ? { value, label: value } : value));

    const onChange = (selectValue) => {
      if (attributes.multi) {
        const values = _.map(selectValue, val => val.value);
        attributes.onChange(values);
      } else if (selectValue !== null && selectValue !== undefined) {
        attributes.onChange(selectValue.value);
      } else {
        attributes.onChange(null);
      }
    };

    const delimiter = attributes.delimiter || ';';

    return (
      <div className="col-md-4">
        <Select
          name={attributes.id}
          {...attributes}
          options={options}
          delimiter={delimiter}
          value={attributes.multi ? _.join(attributes.value, delimiter) : attributes.value}
          onChange={onChange}
        />
      </div>
    );
  };

  return (
    <BaseReactField
      {...props}
      renderInput={renderInput}
    />
  );
};

export default SelectField;
