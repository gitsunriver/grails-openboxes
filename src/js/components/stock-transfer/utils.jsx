import _ from 'lodash';

const PENDING = 'PENDING';
const CANCELED = 'CANCELED';

function extractItem(item, status) {
  const { destinationBinLocation, destinationZone } = item;
  return {
    ...item,
    destinationBinLocation: {
      id: destinationBinLocation && destinationBinLocation.id ? destinationBinLocation.id : '',
      name: destinationBinLocation && destinationBinLocation.name ? destinationBinLocation.name : '',
      zoneId: destinationZone && destinationZone.id ? destinationZone.id : null,
      zoneName: destinationZone && destinationZone.name ? destinationZone.name : null,
    },
    quantity: status === PENDING ? '' : item.quantity,
  };
}

function extractSplitItem(item, splitItem, status) {
  return {
    ...splitItem,
    destinationBinLocation: {
      id: splitItem.destinationBinLocation.id ? splitItem.destinationBinLocation.id : '',
      name: splitItem.destinationBinLocation.name ? splitItem.destinationBinLocation.name : '',
      zoneId: splitItem.destinationBinLocation.zone ?
        splitItem.destinationBinLocation.zone.id : null,
      zoneName: splitItem.destinationBinLocation.zone ?
        splitItem.destinationBinLocation.zone.name : null,
    },
    quantity: status === PENDING ? '' : item.quantity,
    quantityOnHand: item.quantityOnHand,
    referenceId: item.id, // set a referenceId from original item
  };
}

export function extractStockTransferItems(stockTransfer) {
  const stockTransferItems = [];
  _.forEach(stockTransfer.stockTransferItems, (item) => {
    stockTransferItems.push(extractItem(item, stockTransfer.status));
    if (item.splitItems.length > 0) {
      _.forEach(item.splitItems, (splitItem) => {
        stockTransferItems.push(extractSplitItem(item, splitItem, stockTransfer.status));
      });
    }
  });
  return stockTransferItems;
}

export function extractNonCanceledItems(stockTransfer) {
  const stockTransferItems = [];
  _.forEach(stockTransfer.stockTransferItems, (item) => {
    if (item.status !== CANCELED) {
      stockTransferItems.push(extractItem(item, stockTransfer.status));
    }
    if (item.splitItems.length > 0) {
      _.forEach(item.splitItems, (splitItem) => {
        stockTransferItems.push(extractSplitItem(item, splitItem, stockTransfer.status));
      });
    }
  });
  return stockTransferItems;
}

export function prepareRequest(stockTransfer, status) {
  const originalItems = _.filter(stockTransfer.stockTransferItems, item => !item.referenceId);
  const stockTransferItems = _.map(originalItems, item => ({
    ...item,
    splitItems: _.filter(
      stockTransfer.stockTransferItems,
      splitItem => splitItem.referenceId === item.id,
    ),
    quantity: item.quantity === '' ? item.quantityOnHand : item.quantity,
  }));

  return { ...stockTransfer, stockTransferItems, status };
}
