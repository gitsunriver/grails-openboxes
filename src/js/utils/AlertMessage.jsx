import React from 'react';
import PropTypes from 'prop-types';

const AlertMessage = ({
  show, message, danger, close,
}) => {
  if (show) {
    return (
      <div
        className={`alert ${danger ? 'alert-danger' : 'alert-warning'}`}
        style={{ cursor: close ? 'pointer' : 'default' }}
        role="presentation"
        onClick={() => close()}
      >
        <i className={`fa ${danger ? 'fa-times-circle' : 'fa-exclamation-circle'} pr-2`} />
        {message}
      </div>);
  }

  return null;
};

export default AlertMessage;

AlertMessage.propTypes = {
  show: PropTypes.bool,
  message: PropTypes.string,
  danger: PropTypes.bool,
  close: PropTypes.func,
};

AlertMessage.defaultProps = {
  show: false,
  message: '',
  danger: false,
  close: () => null,
};
