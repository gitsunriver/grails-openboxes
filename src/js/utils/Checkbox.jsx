import React from 'react';
import PropTypes from 'prop-types';

const Checkbox = ({
  value, indeterminate, custom, ...props
}) => {
  const onChange = (event) => {
    const { checked } = event.target;

    if (props.onChange) {
      props.onChange(checked);
    }
  };

  if (custom) {
    return (
      <div className="custom-checkbox">
        <input
          type="checkbox"
          checked={value}
          {...props}
          onChange={onChange}
        />
        <label htmlFor={props.id} />
      </div>
    );
  }

  return (
    <input
      type="checkbox"
      // eslint-disable-next-line no-param-reassign,no-return-assign
      ref={elem => elem && (elem.indeterminate = indeterminate)}
      checked={value}
      {...props}
      onChange={onChange}
    />);
};

export default Checkbox;

Checkbox.propTypes = {
  onChange: PropTypes.func,
  value: PropTypes.bool,
  indeterminate: PropTypes.bool,
  custom: PropTypes.bool,
  id: PropTypes.string,
};

Checkbox.defaultProps = {
  onChange: null,
  value: null,
  indeterminate: false,
  custom: false,
  id: '',
};
