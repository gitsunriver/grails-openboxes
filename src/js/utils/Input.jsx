import React from 'react';
import PropTypes from 'prop-types';

const Input = (props) => {
  const onChange = (event) => {
    const { value } = event.target;

    if (props.onChange) {
      props.onChange(value);
    }
  };

  return (
    <input
      type="text"
      {...props}
      onChange={onChange}
    />
  );
};

export default Input;

Input.propTypes = {
  onChange: PropTypes.func,
};

Input.defaultProps = {
  onChange: null,
};
