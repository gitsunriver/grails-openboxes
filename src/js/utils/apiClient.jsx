/* eslint-disable no-console */
import axios from 'axios';

const justRejectRequestError = error => Promise.reject(error);

const apiClient = axios.create({});

const handleSuccess = response => response;

const handleError = (error) => {
  switch (error.response.status) {
    case 400:
      console.error('Bad request.');
      break;
    case 401:
      console.error('Unauthorized.');
      break;
    case 403:
      console.error('Access denied.');
      break;
    case 404:
      console.error('Not found.');
      break;
    case 500:
      console.error('Internal server error.');
      break;
    default: {
      console.error(error);
    }
  }
  return Promise.reject(error);
};

apiClient.interceptors.response.use(handleSuccess, handleError);
apiClient.interceptors.request.use(config => config, justRejectRequestError);

export default apiClient;
