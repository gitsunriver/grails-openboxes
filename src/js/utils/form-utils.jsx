import React from 'react';
import PropTypes from 'prop-types';

export const renderFormField = (fieldConfig, fieldName, props = {}) => {
  const FieldType = fieldConfig.type;

  return (
    <FieldType
      key={fieldName}
      fieldName={fieldName}
      fieldConfig={fieldConfig}
      {...props}
    />
  );
};

export const renderField = ({
  renderInput,
  attributes: { required, hidden, ...otherAttributes },
  label: FieldLabel,
  touched: fieldTouched,
  input,
  meta: { touched, error },
}) => {
  const attr = { id: input.name, ...otherAttributes };

  const className = `form-group ${required ? 'required' : ''} ${hidden ? 'd-none' : ''} ${(touched || fieldTouched) && error ? 'has-error' : ''}`;

  return (
    <div className={`padding-left-md padding-right-md ${className}`}>
      <div className="row">
        {
          typeof FieldLabel === 'string' ?
            <label htmlFor={attr.id} className="col-md-2 col-form-label">{ FieldLabel }</label> :
            <FieldLabel />
        }
        {renderInput(input, attr)}
      </div>
      <div className="row">
        <div className="col-md-2" />
        <div className="help-block col-md-4" style={{ float: 'left' }}>
          { touched || fieldTouched ? error : '' }
        </div>
      </div>
    </div>
  );
};

renderField.propTypes = {
  renderInput: PropTypes.func.isRequired,
  attributes: PropTypes.shape({}).isRequired,
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
  touched: PropTypes.bool,
  input: PropTypes.shape({}).isRequired,
  meta: PropTypes.shape({}).isRequired,
};

renderField.defaultProps = {
  touched: false,
  label: '',
};
