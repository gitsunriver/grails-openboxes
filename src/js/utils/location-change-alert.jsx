import { confirmAlert } from 'react-confirm-alert';

import { dispatch } from '../store';
import { changeCurrentLocation } from '../actions';

const showLocationChangedAlert = (translate, oldLocation, newLocation, onLocationChanged) => {
  if (oldLocation.id && oldLocation.id !== newLocation.id) {
    confirmAlert({
      title: translate('message.locationChanged.label', 'Location was changed'),
      message: translate(
        'locationChanged.message',
        'Current location was changed, do you want to change the location back and continue work? If you press No all not saved changes will be lost.',
      ),
      buttons: [
        {
          label: translate('default.yes.label', 'Yes'),
          onClick: () => { dispatch(changeCurrentLocation(oldLocation)); },
        },
        {
          label: translate('default.no.label', 'No'),
          onClick: () => {
            if (onLocationChanged) {
              onLocationChanged();
            }
          },
        },
      ],
    });
  }
};

export default showLocationChangedAlert;
