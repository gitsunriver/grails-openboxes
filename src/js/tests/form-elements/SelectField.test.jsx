import React from 'react';
import renderer from 'react-test-renderer';
import SelectField from '../../components/form-elements/SelectField';
import { renderFormField } from '../../utils/form-utils';

jest.mock('redux-form', () => ({
  Field: (props) => {
    const { component: Component, name, ...others } = props;
    const input = { onChange: () => {}, name };
    const meta = {};
    return <Component input={input} meta={meta} {...others} />;
  },
}));

describe('SelectField component is correctly rendering', () => {
  it('renders correctly', () => {
    const fieldConfig = {
      type: SelectField,
      label: 'test label',
      attributes: {
        options: ['One', 'Two'],
      },
    };

    const rendered = renderer.create(renderFormField(fieldConfig, 'test-field'));

    expect(rendered.toJSON()).toMatchSnapshot();
  });
});

