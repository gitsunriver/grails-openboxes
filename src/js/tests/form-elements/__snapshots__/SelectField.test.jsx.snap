// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField component is correctly rendering renders correctly 1`] = `
<div
  className="padding-left-md padding-right-md form-group my-0   "
>
  <div
    className="row"
  >
    <label
      className="col-md-2 col-5 col-form-label text-center text-md-right"
      htmlFor="test-field"
    >
      test label
    </label>
    <div
      className="col-md-4"
    >
      <div
        id="select-id_1-container"
      >
        <div
          className="Select col-md-12 col-5 is-clearable is-searchable Select--single"
          style={undefined}
        >
          <div
            className="Select-control"
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={undefined}
          >
            <span
              className="Select-multi-value-wrapper"
              id="react-select-2--value"
            >
              <div
                className="Select-placeholder"
              >
                Select...
              </div>
              <div
                className="Select-input"
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-activedescendant="react-select-2--value"
                  aria-describedby={undefined}
                  aria-expanded="false"
                  aria-haspopup="false"
                  aria-label={undefined}
                  aria-labelledby={undefined}
                  aria-owns=""
                  className={undefined}
                  id="test-field"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  role="combobox"
                  style={
                    Object {
                      "boxSizing": "content-box",
                      "width": "5px",
                    }
                  }
                  tabIndex={undefined}
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </span>
            <span
              className="Select-arrow-zone"
              onMouseDown={[Function]}
            >
              <span
                className="Select-arrow"
                onMouseDown={[Function]}
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-2"
    />
    <div
      className="help-block col-md-4"
      style={
        Object {
          "float": "left",
        }
      }
    >
      
    </div>
  </div>
</div>
`;
